game/gameboard.py

```diff
@@ -42,31 +43,41 @@ def put_chip(self, column_index, color):
             column_index: the column where the chip is being put
             color: the color of the chip
         """
-        if not self.column_is_full(column_index):
-            rowIndex = self.ROWSCOUNT - \
-                list(self._retrieve_column(column_index)).index(None) - 1
-            self._matrix[rowIndex][column_index - 1] = color.lower()
+        if not self.board_is_full():
+            if not self.column_is_full(column_index):
+                rowIndex = self.ROWSCOUNT - \
+                    list(self._retrieve_column(column_index)).index(None) - 1
+                self._matrix[rowIndex][column_index - 1] = color.lower()
+            else:
+                raise ColumnIsFull(column_index)
         else:
-            raise ColumnIsFull(column_index)
+            raise BoardIsFull(self._matrix)
+
 
     def retrieve_first_non_full_column(self):
         """Returns the index of the first non-full column
            found, looking from left to right on the board
         """
-        index_list = range(1, self.COLUMNSCOUNT + 1)
-        for index in index_list:
-            if not self.column_is_full(index):
-                return index
+        if not self.board_is_full():
+            index_list = range(1, self.COLUMNSCOUNT + 1)
+            for index in index_list:
+                if not self.column_is_full(index):
+                    return index
+        else:
+            raise BoardIsFull(self._matrix)
 
     def retrieve_emptiest_column(self):
         """Returns the index of the emptiest column found on the board
         """
-        rows_index_list = range(1, self.ROWSCOUNT + 1)
-        columns_index_list = range(1, self.COLUMNSCOUNT + 1)
-        for row_index in rows_index_list:
-            for column_index in columns_index_list:
-                if self.read_entry(row_index, column_index) is None:
-                    return column_index
+        if not self.board_is_full():
+            rows_index_list = range(1, self.ROWSCOUNT + 1)
+            columns_index_list = range(1, self.COLUMNSCOUNT + 1)
+            for row_index in rows_index_list:
+                for column_index in columns_index_list:
+                    if self.read_entry(row_index, column_index) is None:
+                        return column_index
+        else:
+            raise BoardIsFull(self._matrix)
```

Comment:
Very minor change, but interesting to see how doing a slight change in the programming style can help you to factor out duplicated code. Consider this last change, it could be written as:

```
    def retrieve_emptiest_column(self):
       if self.board_is_full():
            raise BoardIsFull(self._matrix)

        rows_index_list = range(1, self.ROWSCOUNT + 1)
        columns_index_list = range(1, self.COLUMNSCOUNT + 1)
        for row_index in rows_index_list:
            for column_index in columns_index_list:
                if self.read_entry(row_index, column_index) is None:
                    return column_index
 ```

now, that same thing can be applied to the previos methods where you do a similar check. But hey, now you have a repeated piece of code that you can extract! So now you can do:

```
def _raise_if_board_full(self):
       if self.board_is_full():
            raise BoardIsFull(self._matrix)

```
    def retrieve_emptiest_column(self):
        self._raise_if_board_is_full()

        rows_index_list = range(1, self.ROWSCOUNT + 1)
        columns_index_list = range(1, self.COLUMNSCOUNT + 1)
        for row_index in rows_index_list:
            for column_index in columns_index_list:
                if self.read_entry(row_index, column_index) is None:
                    return column_index
 ```